### PR TITLE
feat: Using multi_set in shell tool to speedup data copy

### PR DESCRIPTION
**Pull request by Arsenal821**
### What problem does this PR solve? 
https://github.com/apache/incubator-pegasus/issues/923
pegasus_shell copy_data use async_set write now, I want use async_multi_set improve speed sometimes

### What is changed and how does it work?
You can use pegasus_shell exec "copy_data -c cluster_name -a app_name -u"  for it

### The performance tests between async_set and async_multi_set for copy_data:
The same row show different cost from copy the same table to a new empty table. 
|  total_data_count   | row data pre hash_key | async_set (secs) | async_multi_set (secs)  |
|  ----  | ----  |----  | ----  |
| 152,354 | 1   | 8 | 16 |
| 417,408 | 4   | 29 | 11 |
| 657,950 | 50   | 32 | 9 |
| 2,079,400 | 200   | 87 | 18 |
| 1,842,400 | 800   | 399 | 19 |
| 14,374,001 | 2,000   | 631 | 131 |

You should use async_multi_set for copy when your data have multi count per hash_key.

### CODE REVIEW

**Comment by acelyc111** [comment]
Good work!
Could you plz describe how faster gain after this patch in your test enviroment?

File: src/shell/command_helper.h [hunk 1]
```diff
@@ -36,6 +36,7 @@
 #include <dsn/perf_counter/perf_counter_utils.h>
 #include <dsn/utility/string_view.h>
 #include <dsn/utils/time_utils.h>
+#include <dsn/utility/synchronize.h>
```
Comment (acelyc111) [inline src/shell/command_helper.h:39]: Keep these headers in lexicographical order.

File: src/shell/command_helper.h [hunk 2]
```diff
@@ -271,6 +283,82 @@ inline int compute_ttl_seconds(uint32_t expire_ts_seconds, bool &ts_expired)
     return 0;
 }
 
+inline void batch_execute_multi_set(scan_data_context *context)
+{
+    for (auto it = context->multi_kvs.begin(); it != context->multi_kvs.end(); ++it) {
+        // wait for satisfied with  max_multi_set_concurrency
```
Comment (acelyc111) [inline src/shell/command_helper.h:289]:
```suggestion
        // wait for satisfied with max_multi_set_concurrency
```

File: src/shell/command_helper.h [hunk 3]
```diff
@@ -271,6 +283,82 @@ inline int compute_ttl_seconds(uint32_t expire_ts_seconds, bool &ts_expired)
     return 0;
 }
 
+inline void batch_execute_multi_set(scan_data_context *context)
+{
+    for (auto it = context->multi_kvs.begin(); it != context->multi_kvs.end(); ++it) {
```
Comment (acelyc111) [inline src/shell/command_helper.h:288]:
```suggestion
    for (const auto& it : context->multi_kvs) {
```

File: src/shell/commands/data_operations.cpp [hunk 4]
```diff
@@ -1676,13 +1694,28 @@ bool copy_data(command_executor *e, shell_context *sc, arguments args)
         return false;
     }
 
+    if (max_multi_set_concurrency < 1) {
```
Comment (acelyc111) [inline src/shell/commands/data_operations.cpp:1697]: Should return false when `max_multi_set_concurrency <= 1`?

File: src/shell/command_helper.h (hunk 3 above)
Comment (levy5307) [reply src/shell/command_helper.h:288]: for (const auto& kv : context->multi_kvs) {

File: src/shell/command_helper.h [hunk 5]
```diff
@@ -36,6 +36,7 @@
 #include <dsn/perf_counter/perf_counter_utils.h>
 #include <dsn/utility/string_view.h>
 #include <dsn/utils/time_utils.h>
```
Comment (levy5307) [inline src/shell/command_helper.h:38]: Could you add some performance tests for this pull request?

**Comment by Arsenal821** [comment]
The performance tests between async_set and async_multi_set for copy_data:
The same row show different cost from copy the same table to a new empty table. 
|  case   | seconds for async_set | seconds for async_multi_set  |
|  ----  | ----  | ----  |
| total_data_count: 152354, 1 row data pre hash_key  | 8 | 16 |
| total_data_count: 417408, 4 rows data pre hash_key  | 29 | 11 |
| total_data_count: 657950, 50 rows data pre hash_key  | 32 | 9 |
| total_data_count: 2079400, 200 rows data pre hash_key  | 87 | 18 |
| total_data_count: 1842400, 800 rows data pre hash_key  | 399 | 19 |
| total_data_count: 14374001, 2000 rows data pre hash_key  | 631 | 131 |

You should use async_multi_set for copy when your data have multi count per hash_key.

**Comment by Arsenal821** [comment]
https://github.com/apache/incubator-pegasus/issues/923

File: src/shell/command_helper.h (hunk 5 above)
Comment (Arsenal821) [reply src/shell/command_helper.h]: done

File: src/shell/command_helper.h [hunk 6]
```diff
@@ -271,6 +283,83 @@ inline int compute_ttl_seconds(uint32_t expire_ts_seconds, bool &ts_expired)
...
+                            context->multi_kvs.emplace(hash_key,
+                                                       std::map<std::string, std::string>());
+                        }
+                        if (expire_ts_seconds && context->multi_expire_ts_seconds < ttl_seconds) {
+                            context->multi_expire_ts_seconds = expire_ts_seconds;
+                        }
+                        context->multi_kvs[hash_key].emplace(std::move(sort_key), std::move(value));
+
+                        if (context->data_count >= context->max_batch_count) {
+                            batch_execute_multi_set(context);
```
[the remark this answers is not in the document]
Comment (Arsenal821) [reply src/shell/command_helper.h:342]: if do it, we will lose multi_set preponderance if ttl time have most difference. In a compromise, our business chose to do so.  Could U think I must do it?

[the remark this answers is not in the document]
Comment (acelyc111) [reply src/shell/command_helper.h:342]: We can implement it in the future, add an issue to track it, maybe we have to add an new API, like multiSet on different sort keys with different TTLs.
Now, better to add a warnning notice in shell when using it.

[the remark this answers is not in the document]
Comment (Shuo-Jia) [reply src/shell/command_helper.h:342]: > We can implement it in the future, add an issue to track it, maybe we have to add a new API, like multiSet on different sort keys with different TTLs.
> Now, better to add a warnning notice in shell when using it.

OK！

[the remark this answers is not in the document]
Comment (Arsenal821) [reply src/shell/command_helper.h:342]: I have done

File: src/shell/command_helper.h [hunk 7]
```diff
@@ -271,6 +283,83 @@ inline int compute_ttl_seconds(uint32_t expire_ts_seconds, bool &ts_expired)
...
+                    int ttl_seconds = 0;
+                    ttl_seconds = compute_ttl_seconds(expire_ts_seconds, ts_expired);
+                    if (!ts_expired) {
+                        context->data_count++;
+                        if (context->multi_kvs.find(hash_key) == context->multi_kvs.end()) {
+                            context->multi_kvs.emplace(hash_key,
+                                                       std::map<std::string, std::string>());
+                        }
+                        if (expire_ts_seconds && context->multi_expire_ts_seconds < ttl_seconds) {
+                            context->multi_expire_ts_seconds = expire_ts_seconds;
```
[the remark this answers is not in the document]
Comment (Arsenal821) [reply src/shell/command_helper.h]: improve done

**Review by acelyc111 — APPROVED** [review]
Good work! Thanks for your contribution.

**Review by Shuo-Jia — APPROVED** [review]

**Pull request merged**